### PR TITLE
Improve MySQL compatibility of statement `SHOW INDEX`

### LINKED ISSUE #51782: MySQL compatibility: SHOW KEYS output
**Describe the unexpected behaviour**

BI tools such as QuickSight use `SHOW KEYS` queries to determine the primary keys of the dataset.
The output of the `SHOW KEYS` command is different from a real MySQL server.

"Real" MySQL:

```
mysql> SHOW KEYS FROM `cell_towers` FROM `mysql`;
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+------------+
| Table       | Non_unique | Key_name | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment | Visible | Expression |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+------------+
| cell_towers |          0 | PRIMARY  |            1 | radio       | A         |           0 |     NULL |   NULL |      | BTREE      |         |               | YES     | NULL       |
| cell_towers |          0 | PRIMARY  |            2 | mcc         | A         |           0 |     NULL |   NULL |      | BTREE      |         |               | YES     | NULL       |
| cell_towers |          0 | PRIMARY  |            3 | net         | A         |           0 |     NULL |   NULL |      | BTREE      |         |               | YES     | NULL       |
| cell_towers |          0 | PRIMARY  |            4 | created     | A         |           0 |     NULL |   NULL |      | BTREE      |         |               | YES     | NULL       |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+------------+
```

ClickHouse via MySQL interface:

```
mysql> SHOW KEYS FROM `cell_towers` FROM `default`;
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+--------------------------+
| table       | non_unique | key_name | seq_in_index | column_name | collation | cardinality | sub_part | packed | null | index_type | comment | index_comment | visible | expression               |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+--------------------------+
| cell_towers |          0 | PRIMARY  | NULL         | NULL        | A         | NULL        | NULL     | NULL   | NULL | primary    | NULL    | NULL          | YES     | radio, mcc, net, created |
+-------------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+--------------------------+
```

**How to reproduce**
* Which ClickHouse server version to use: 23.6
* Which interface to use, if it matters: MySQL
* Non-default settings, if any: I added `use_mysql_types_in_show_columns` to the default profile.

```xml
  <profiles>
    <default>
      <load_balancing>random</load_balancing>
      <use_mysql_types_in_show_columns>1</use_mysql_types_in_show_columns>
    </default>
  </profiles>
```

* `CREATE TABLE` statements for all tables involved: [cell towers dataset](https://clickhouse.com/docs/en/getting-started/example-datasets/cell-towers) for ClickHouse, and a simple table like
```
create table cell_towers (radio TEXT, mcc INTEGER, net INTEGER, created DATETIME, PRIMARY KEY (radio, mcc, net, created));
```
for MySQL

* Queries to run that lead to unexpected result

```
SHOW KEYS FROM `cell_towers` FROM `default`;
```

**Expected behavior**
* The primary key is reported as a single row, whereas a composite key in MySQL is reported as 1 row per column
* There are NULLs in the output instead of numbers and/or empty strings in a real MySQL, and that supposedly causes QuickSight errors.

CC @tpanetti

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #51782

This PR improves MySQL compatibility of statement `SHOW INDEX` as follows:
- fields `seq_in_index`, `column_name` and `expression` now correspond better to MySQLs behavior. There are some caveats documented in the code (fixing these is only possible with much more sophisticated SQL or rewriting `SHOW INDEX` in C++)
- field `non_unique` now always shows `1` instead of `0`. This is because ClickHouse does not support uniqueness constraints (like MySQL).
- field `cardinality` (an estimation) now always shows `0` instead of `NULL`. The true cardinality cannot be determined cheaply. (Even in MySQL the cardinality calculation must be triggered explicitly and only the last cached value is shown.)
- field `sub_part` is now documented to be always `NULL` instead of being "reserved" . This is compatible with MySQL.
- field `packed` is now documented to be always `NULL` instead of being "reserved". This is compatible with MySQL.
- field `index_type` now shows an uppercase index type, e.g. `PRIMARY` instead of `primary`.
- fields `comment` and `index_comment` now produce '' (empty string) instead of `NULL`.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve MySQL compatibility of statement SHOW INDEX

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 4da14e2be02a3d01e0e73f446188ea9223be7851 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51796/4da14e2be02a3d01e0e73f446188ea9223be7851/ci_running.html)
The overall status of the commit is 🟢 success

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟢 success</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: docs/en/sql-reference/statements/show.md [hunk 1]
```diff
@@ -281,22 +285,22 @@ equivalent. If no database is specified, the query assumes the current database
...
-- column_name - Currently unused
-- collation - The sorting of the column in the index, `A` if ascending, `D` if descending, `NULL` if unsorted (Nullable(String))
-- cardinality - Currently unused
-- sub_part - Currently unused
-- packed - Currently unused
+The statement produces a result table with the following structure:
+- table - The name of the table. (String)
+- non_unique - Always `1` as ClickHouse does not support uniqueness constraints. (UInt8)
+- key_name - The name of the index, `PRIMARY` if the index is a primary key index. (String)
+- column_name - For a primary key index, the name of the column. For a data skipping index: '' (empty string), see field "expression". (String)
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/statements/show.md:292]: Done

File: docs/en/sql-reference/statements/show.md [hunk 2]
```diff
@@ -281,22 +285,22 @@ equivalent. If no database is specified, the query assumes the current database
...
+- packed - Always `NULL` because ClickHouse does not support packed indexes (like MySQL). (Nullable(String))
 - null - Currently unused
-- index_type - The index type, e.g. `primary`, `minmax`, `bloom_filter` etc. (String)
-- comment - Currently unused
-- index_comment - Currently unused
-- visible - If the index is visible to the optimizer, always `YES` (String)
-- expression - The index expression (String)
+- index_type - The index type, e.g. `PRIMARY`, `MINMAX`, `BLOOM_FILTER` etc. (String)
+- comment - Additional information about the index, currently always `` (empty string). (String)
+- index_comment - `` (empty string) because indexes in ClickHouse cannot have a `COMMENT` field (like in MySQL). (String)
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/statements/show.md:301]: Done

File: docs/en/sql-reference/statements/show.md [hunk 3]
```diff
@@ -281,22 +285,22 @@ equivalent. If no database is specified, the query assumes the current database
...
+- sub_part - Always `NULL` because ClickHouse does not support index prefixes like MySQL. (Nullable(String))
+- packed - Always `NULL` because ClickHouse does not support packed indexes (like MySQL). (Nullable(String))
 - null - Currently unused
-- index_type - The index type, e.g. `primary`, `minmax`, `bloom_filter` etc. (String)
-- comment - Currently unused
-- index_comment - Currently unused
-- visible - If the index is visible to the optimizer, always `YES` (String)
-- expression - The index expression (String)
+- index_type - The index type, e.g. `PRIMARY`, `MINMAX`, `BLOOM_FILTER` etc. (String)
+- comment - Additional information about the index, currently always `` (empty string). (String)
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/statements/show.md:300]: Done

File: src/Interpreters/InterpreterShowIndexesQuery.cpp [hunk 4]
```diff
@@ -40,52 +40,67 @@ SELECT *
 FROM (
         (SELECT
             name AS table,
-            0 AS non_unique,
+            1 AS non_unique,
             'PRIMARY' AS key_name,
-            NULL AS seq_in_index,
-            NULL AS column_name,
+            arrayJoin(splitByString(', ', primary_key)) AS column_name,
+            row_number() over (order by column_name) AS seq_in_index,
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Interpreters/InterpreterShowIndexesQuery.cpp:46]: I thought that the `ORDER BY column_name` needs to be preceded by `column_name`. Anyways, it works also in the original form, I change it back.

**Comment by rschu1ze** [comment]
Restarted some failing tests.

**Comment by rschu1ze** [comment]
@novikd Approve?

**Review by novikd — APPROVED** [review]

File: src/Interpreters/InterpreterShowIndexesQuery.cpp [hunk 5]
```diff
@@ -40,52 +40,67 @@ SELECT *
 FROM (
         (SELECT
             name AS table,
-            0 AS non_unique,
+            1 AS non_unique,
             'PRIMARY' AS key_name,
-            NULL AS seq_in_index,
-            NULL AS column_name,
+            row_number() over (order by column_name) AS seq_in_index,
+            arrayJoin(splitByString(', ', primary_key)) AS column_name,
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowIndexesQuery.cpp:46]: Dirty, but passable.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**